src/_macosx.m

```diff
@@ -1748,19 +1771,16 @@ - (void)flagsChanged:(NSEvent *)event
     }
 
     // hold a reference to the timer so we can invalidate/stop it later
-    self->timer = [NSTimer timerWithTimeInterval: interval
-                                         repeats: !single
-                                           block: ^(NSTimer *timer) {
-        gil_call_method((PyObject*)self, "_on_timer");
-        if (single) {
-            // A single-shot timer will be automatically invalidated when it fires, so
-            // we shouldn't do it ourselves when the object is deleted.
-            self->timer = NULL;
-        }
+    self->timer = [NSTimer scheduledTimerWithTimeInterval: interval
+                                                  repeats: !single
+                                                    block: ^(NSTimer *timer) {
+        dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
This is the major change here. We are dispatching our event asynchronously (so as to not sequentially depend on the callback processing time), but putting it onto the main thread's queue.